go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -42,6 +43,48 @@ func addService(service *kapi.Service, inport, outport, gwBridge string, nodeIP
 				klog.Errorf("Failed to add openflow flow on %s for nodePort: "+
 					"%d, stderr: %q, error: %v", gwBridge, svcPort.NodePort, stderr, err)
 			}
+			// Flows for cloud load balancers on Azure/GCP
+			flows := make([]string, 0)
+
+			// Established traffic is handled by default conntrack rules
+			// Table 1 handles forwarding traffic to pods
+			// Table 2 handles return NAT traffic for node port and forwards out of the host
+			// NodePort/Ingress access in the OVS bridge will only ever come from outside of the host
+			for _, ing := range service.Status.LoadBalancer.Ingress {
+				if ing.IP == "" {
+					continue
+				}
+				ingIP := net.ParseIP(ing.IP)
+				if ingIP == nil {
+					klog.Errorf("Failed to parse ingress IP: %s", ing.IP)
+					continue
+				}
+				cookie, err := svcToCookie(service.Name, ingIP, svcPort.Port)
+				if err != nil {
+					klog.Errorf("Unable to generate cookie for svc: %s, %s, %d, error: %v",
+						service.Name, ingIP.String(), svcPort.Port, err)
+				}
+				ingPort := fmt.Sprintf("%d", svcPort.Port)
+				// Send to CT from external net with DNAT, eventually will land in table 1 (traffic to pods)
+				flows = append(flows,
+					fmt.Sprintf("cookie=%s,priority=101,in_port=%s,ct_state=-trk,%s,nw_dst=%s,tp_dst=%s,"+
+						"actions=ct(commit,table=0,zone=%d,nat(dst=%s)",
+						cookie, inport, protocol, ing.IP, ingPort, config.Default.ConntrackZone,
+						util.JoinHostPortInt32(nodeIP.IP.String(), svcPort.NodePort)))
+				// Incoming return traffic from pods, send to CT, unDNAT, goto table 2
```

Comment:
traffic coming from the pods might not be going out of the host, it might be going to the host as well. So I needed to differentiate that by matching on the particular nodeport. Since no host-originated nodeport traffic will ever go through breth0, we dont have to worry about hijacking that traffic in breth0 and can send directly to table 2.